pkg/scrape/discovery/file/file.go

```diff
@@ -0,0 +1,418 @@
+
+// TimestampCollector is a Custom Collector for Timestamps of the files.
+type TimestampCollector struct {
+	Description *prometheus.Desc
+	discoverers map[*Discovery]struct{}
+	lock        sync.RWMutex
+}
+
+// Describe method sends the description to the channel.
+func (t *TimestampCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- t.Description
+}
+
+// Collect creates constant metrics for each file with last modified time of the file.
+func (t *TimestampCollector) Collect(ch chan<- prometheus.Metric) {
+	// New map to dedup filenames.
+	uniqueFiles := make(map[string]float64)
+	t.lock.RLock()
+	for fileSD := range t.discoverers {
+		fileSD.lock.RLock()
+		for filename, timestamp := range fileSD.timestamps {
+			uniqueFiles[filename] = timestamp
+		}
+		fileSD.lock.RUnlock()
+	}
+	t.lock.RUnlock()
+	for filename, timestamp := range uniqueFiles {
+		ch <- prometheus.MustNewConstMetric(
+			t.Description,
+			prometheus.GaugeValue,
+			timestamp,
+			filename,
+		)
+	}
+}
+
+func (t *TimestampCollector) addDiscoverer(disc *Discovery) {
+	t.lock.Lock()
+	t.discoverers[disc] = struct{}{}
+	t.lock.Unlock()
+}
+
+func (t *TimestampCollector) removeDiscoverer(disc *Discovery) {
+	t.lock.Lock()
+	delete(t.discoverers, disc)
+	t.lock.Unlock()
+}
+
+// NewTimestampCollector creates a TimestampCollector.
+func NewTimestampCollector() *TimestampCollector {
+	return &TimestampCollector{
+		Description: prometheus.NewDesc(
+			"prometheus_sd_file_mtime_seconds",
+			"Timestamp (mtime) of files read by FileSD. Timestamp is set at read time.",
+			[]string{"filename"},
+			nil,
+		),
+		discoverers: make(map[*Discovery]struct{}),
+	}
+}
+
+// Discovery provides service discovery functionality based
+// on files that contain target groups in JSON or YAML format. Refreshing
+// happens using file watches and periodic refreshes.
+type Discovery struct {
+	paths      []string
+	watcher    *fsnotify.Watcher
+	interval   time.Duration
+	timestamps map[string]float64
+	lock       sync.RWMutex
+
+	// lastRefresh stores which files were found during the last refresh
+	// and how many target groups they contained.
+	// This is used to detect deleted target groups.
+	lastRefresh map[string]int
+	logger      logrus.FieldLogger
+}
+
+// NewDiscovery returns a new file discovery for the given paths.
+func NewDiscovery(conf *SDConfig, logger logrus.FieldLogger) *Discovery {
+
+	disc := &Discovery{
+		paths:      conf.Files,
+		interval:   time.Duration(conf.RefreshInterval),
+		timestamps: make(map[string]float64),
+		logger:     logger,
+	}
+	fileSDTimeStamp.addDiscoverer(disc)
+	return disc
+}
+
+// listFiles returns a list of all files that match the configured patterns.
+func (d *Discovery) listFiles() []string {
+	var paths []string
+	for _, p := range d.paths {
+		files, err := filepath.Glob(p)
+		if err != nil {
+			d.logger.WithError(err).Error("msg", "Error expanding glob", "glob", p, "err", err)
```

Comment:
Oh, if I'm not mistaken all the fields should be specified explicitly:
```suggestion
			d.logger.WithError(err).WithField("glob", p).Error("expanding glob")
```

I think other places should be amended accordingly.